### PR TITLE
Optimization of JavaScript Tarball Downloading

### PR DESCRIPTION
**Pull request by fepas**
CLOUDBLD-9007

Signed-off-by: Felipe Campos <fdealmei@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- n/a OpenAPI schema is updated (if applicable)
- n/a DB schema change has corresponding DB migration (if applicable)
- [x] README updated (if worker configuration changed, or if applicable)

### CODE REVIEW

**Comment by fepas** [comment]
@MartinBasti 

image


After researching a lot about the approach of workers in our use case, I ended up finding [this](https://stackoverflow.com/a/48484593) answer from StackOverflow which seemed to me to be the most detailed in relation to the use of semaphore vs workers, in it, it is said:

    ..the idiomatic way of limiting the number of parallel tasks this with asyncio is using asyncio.Semaphore


But this also helped me to understand that `sleep` was not necessary, just using the semaphore as a context manager is enough (removing `sleep` also brought a considerable performance gain).

**Comment by MartinBasti** [comment]
> @MartinBasti
> 
> image
> 
> After researching a lot about the approach of workers in our use case, I ended up finding [this](https://stackoverflow.com/a/48484593) answer from StackOverflow which seemed to me to be the most detailed in relation to the use of semaphore vs workers, in it, it is said:
> 
> ```
> ..the idiomatic way of limiting the number of parallel tasks this with asyncio is using asyncio.Semaphore
> ```
> 
> But this also helped me to understand that `sleep` was not necessary, just using the semaphore as a context manager is enough (removing `sleep` also brought a considerable performance gain).

I don't like idea, that python have to create several thousands of asyncio objects in memory (and also handle them in parallel).
From your link, `...in some cases this approach can have advantages over a semaphore, specifically when the work to be done is very large...` 

I would like to limit concurency at creation of tasks.

File: cachito/workers/pkg_managers/general_js.py [hunk 1]
```diff
@@ -44,6 +51,67 @@
...
+    :rtype: str, str
+    """
+    full_pkg_name, version = dep_identifier.rsplit("@", 1)
+    if "/" in full_pkg_name:
+        short_pkg_name = full_pkg_name.rsplit("/", 1)[1]
+    else:
+        short_pkg_name = full_pkg_name
+
+    proxied_url = f"{proxy_repo_url}{full_pkg_name}/-/{short_pkg_name}-{version}.tgz"
+    tarball_name = f"{full_pkg_name}-{version}.tgz".replace("@", "").replace("/", "-")
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general_js.py:70]: How did you determine the expected format of the url and the downloaded tarball name?

File: cachito/workers/pkg_managers/general_js.py [hunk 2]
```diff
@@ -44,6 +51,67 @@
...
+    async with aiohttp.ClientSession() as session:
+
+        tasks = []
+        for dep_identifier in deps_to_download:
+            proxied_url, tarball_name = parse_dependency(proxy_repo_url, dep_identifier)
+            tasks.append(
+                asyncio.ensure_future(
+                    async_download_binary_file(
+                        session,
+                        proxied_url,
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general_js.py:102]: Very late realization on my part, but I'm not sure if Nexus will have the metadata `npm install` needs if we just download the tarball directly.

I think that for python, Nexus caches both the HTTP response from `/simple/$package_name/` and the tarball itself. I assume something similar happens with npm.

Was this ever verified during the design? I'm thinking how to check it. Maybe just try to install from the outputs of a request

File: cachito/workers/pkg_managers/general_js.py (hunk 1 above)
Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py:70]: https://docs.google.com/document/d/1vG5_GlS_P5f7m87ezqtEnXTwv87rn7jy7yZk6LBU7HA/edit#heading=h.suij2g8c5kf8

It was present in the design document.

File: cachito/workers/pkg_managers/general_js.py (hunk 2 above)
Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py:102]: I don't think this factor was considered at any point in the design. How can I do this test?

Comment (chmeliik) [reply cachito/workers/pkg_managers/general_js.py:102]: Some documentation is in https://github.com/containerbuildsystem/cachito/blob/master/docs/using_requests_locally.md, but when I tried with an integration test scenario, the local install didn't work even with the master branch...

The most reliable way to test it would probably be OSBS integration tests - after deploying to stage :sweat_smile:

Comment (chmeliik) [reply cachito/workers/pkg_managers/general_js.py:102]: So I dove a bit deeper into this and I don't think it should be an issue.

After hunting down the relevant part of the npm CLI code, I found something called a "packument": https://github.com/npm/pacote/blob/main/lib/registry.js#L38. I believe that's the metadata npm needs.

For example, for `chai` (one of the deps in our integration tests), the packument would be this: https://registry.npmjs.org/chai

So I cleaned up my local env, ran your updated version, submitted a Cachito request and checked if the packument was there:
```shell
$ grep -R --files-with-matches "BDD/TDD assertion library for node.js and the browser. Test framework agnostic." tmp/nexus-data/blobs
tmp/nexus-data/blobs/cachito-js/content/vol-22/chap-03/c61d5874-d4a9-4460-9216-a0a2f375d7b4.bytes
tmp/nexus-data/blobs/cachito-js/content/vol-15/chap-03/f89d530d-ff9f-4f57-ae9e-f2753d3a2243.bytes
tmp/nexus-data/blobs/cachito-js/content/vol-33/chap-16/c1490d97-6d61-4320-8eb8-8f519a507a79.bytes

$ jq < tmp/nexus-data/blobs/cachito-js/content/vol-33/chap-16/c1490d97-6d61-4320-8eb8-8f519a507a79.bytes
# indeed that's the packument!
```
...in fact, the only notable difference between the master branch and this branch (as far as the blobs are concerned) is that the master branch downloads 3 copies of the [npm packument](https://registry.npmjs.org/npm) for no reason (and it's massive, something like 19 Mi each). So I think we have all the metadata we need and even get rid of unwanted extras

File: cachito/workers/pkg_managers/general_js.py (hunk 1 above)
Comment (chmeliik) [reply cachito/workers/pkg_managers/general_js.py:70]: While diving through the npm CLI, I also found the relevant code for the filename: https://github.com/npm/cli/blob/84598fe2b84d57b812f53ecc0beba9a04e75c069/workspaces/libnpmpack/lib/index.js#L41

One minor detail is that the '@' should only be replaced if it's at the beginning of the package name, otherwise this seems correct :+1:

File: cachito/workers/pkg_managers/general_js.py (hunk 2 above)
Comment (chmeliik) [reply cachito/workers/pkg_managers/general_js.py:102]: > 3 copies of the [npm packument](https://registry.npmjs.org/npm) for no reason

You know what it could be? The `npm pack` call probably checks for updates and has to download the npm packument to do that. Let's assume missing it is OK, it should be fairly easy to fix if we do need it

Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py:113]: > > @MartinBasti
> > image
> > After researching a lot about the approach of workers in our use case, I ended up finding [this](https://stackoverflow.com/a/48484593) answer from StackOverflow which seemed to me to be the most detailed in relation to the use of semaphore vs workers, in it, it is said:
> > ```
> > ..the idiomatic way of limiting the number of parallel tasks this with asyncio is using asyncio.Semaphore
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But this also helped me to understand that `sleep` was not necessary, just using the semaphore as a context manager is enough (removing `sleep` also brought a considerable performance gain).
> 
> I don't like idea, that python have to create several thousands of asyncio objects in memory (and also handle them in parallel). From your link, `...in some cases this approach can have advantages over a semaphore, specifically when the work to be done is very large...`
> 
> I would like to limit concurency at creation of tasks.

Got it. 
I replaced `asyncio.Semaphore` with a solution where we limit the creation of objects and the number of tasks (using `return_when=asyncio.FIRST_COMPLETED`). This seems to me a very clean solution that does exactly what it needs to do.

File: cachito/workers/pkg_managers/general_js.py (hunk 1 above)
Comment (fepas) [reply cachito/workers/pkg_managers/general_js.py]: Well observed, Adam! I changed it to a regex that checks if the '@' is at the beginning.

File: cachito/workers/pkg_managers/general.py [hunk 3]
```diff
@@ -157,6 +157,36 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
...
+):
+    """
+    Download a binary file (such as a TAR archive) from a URL using asyncio.
+
+    :param aiohttp.ClientSession session: Aiohttp interface for making HTTP requests.
+    :param str url: URL for file download
+    :param str download_dir: Path to download file to
+    :param str tarball_name: Name of the file
+    :param aiohttp.BasicAuth auth: Authentication for the URL
+    :param int chunk_size: Chunk size param for Response.content.read()
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general.py:171]: We try to use type annotations for all new code, could you move the types from the docstring to the definition?

(same for the other new functions)

File: cachito/workers/pkg_managers/general.py [hunk 4]
```diff
@@ -157,6 +157,36 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
...
+    :raise NetworkError: If download failed
+    """
+    # async with concurrency_limit:
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general.py:174]: leftover comment

File: cachito/workers/pkg_managers/general_js.py [hunk 5]
```diff
@@ -44,6 +51,77 @@
...
+    :return: proxied_url for the download and the tarball_name.
+    :rtype: str, str
+    """
+    full_pkg_name, version = dep_identifier.rsplit("@", 1)
+    if "/" in full_pkg_name:
+        short_pkg_name = full_pkg_name.rsplit("/", 1)[1]
+    else:
+        short_pkg_name = full_pkg_name
+
+    proxied_url = f"{proxy_repo_url}{full_pkg_name}/-/{short_pkg_name}-{version}.tgz"
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general_js.py:69]: Minor change to make sure this doesn't break when proxy_repo_url does not end with a `/`
```suggestion
    proxied_url = f"{proxy_repo_url.rstrip('/')}/{full_pkg_name}/-/{short_pkg_name}-{version}.tgz"
```

File: cachito/workers/pkg_managers/general_js.py [hunk 6]
```diff
@@ -44,6 +51,77 @@
...
+    tarball_name = f"{full_pkg_name}-{version}.tgz"
+    # asserts position at start of the string matches the character "@"
+    tarball_name = re.sub(r"^@", "", tarball_name)
+    # matches the character "/"
+    tarball_name = re.sub(r"\/", "-", tarball_name)
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general_js.py:74]: Would it be simpler like this?
```suggestion
    tarball_name = f"{full_pkg_name}-{version}.tgz".removeprefix("@").replace("/", "-")
```
And another suggestion: could you add a comment with a link to the npm code (mentioned in another comment) that explains why we do these replacements?

File: cachito/workers/pkg_managers/general_js.py [hunk 7]
```diff
@@ -138,57 +204,51 @@ def download_dependencies(
...
-                        dir_path = os.path.join(groups["host"], *groups["repo_path"].split("/"))
-                    else:
-                        dir_path = f"external-{dir_path}"
-
-                # Create the target directory for the dependency
-                dep_dir = download_dir.joinpath(*dir_path.split("/", 1))
-                dep_dir.mkdir(exist_ok=True, parents=True)
-                # Move the dependency into the target directory
-                shutil.move(str(download_dir.joinpath(tarball)), str(dep_dir.joinpath(tarball)))
+        # results = asyncio.run(get_dependencies(proxy_repo_url, download_dir, dep_identifiers))
```
Comment (chmeliik) [inline cachito/workers/pkg_managers/general_js.py:221]: leftover comment

File: requirements-test.in [hunk 8]
```diff
@@ -5,4 +5,4 @@ importlib-resources==5.4.0
 jsonschema
 pytest
 pytest-cov
-pyyaml
+pyyaml
```
Comment (chmeliik) [inline requirements-test.in:8]: unnecessary change

File: tests/test_workers/test_pkg_managers/test_general_js.py [hunk 9]
```diff
@@ -116,40 +111,49 @@ def test_download_dependencies(
...
+    # dep_1
+    dep1_source_path = os.path.join(
+        download_dir,
+        "angular-devkit-architect-0.803.26.tgz",
+    )
     dep1_dest_path = os.path.join(
-        download_dir, "@angular-devkit/architect/angular-devkit-architect-0.803.26.tgz"
+        download_dir,
+        "@angular-devkit/architect/angular-devkit-architect-0.803.26.tgz",
+    )
```
Comment (chmeliik) [inline tests/test_workers/test_pkg_managers/test_general_js.py:123]: Did something change in the source/dest paths? If this is just a formatting change, could it be done in a separate commit?

**Review by chmeliik — COMMENTED** [review]
Mostly LGTM, just style / consistency comments

+ could we have unit tests for the new functions?

**Comment by fepas** [comment]
@lkolacek @chmeliik Can you guys check it again? :D

File: tests/test_workers/test_pkg_managers/test_general_js.py [hunk 10]
```diff
@@ -114,7 +114,8 @@ def test_download_dependencies(
 
     dep1_source_path = os.path.join(download_dir, "angular-devkit-architect-0.803.26.tgz")
     dep1_dest_path = os.path.join(
-        download_dir, "@angular-devkit/architect/angular-devkit-architect-0.803.26.tgz",
+        download_dir,
+        "@angular-devkit/architect/angular-devkit-architect-0.803.26.tgz",
```
Comment (chmeliik) [inline tests/test_workers/test_pkg_managers/test_general_js.py:118]: Well now we have one formatting change in this commit and the rest in the previous one :smile: 

If `black` asked for this change it's OK to include it in the same commit, I just didn't understand why black did that (since the code didn't change)

File: tests/test_workers/test_pkg_managers/test_general_js.py [hunk 11]
```diff
@@ -208,23 +192,108 @@ def test_download_dependencies_skip_deps(
...
+@pytest.mark.parametrize(
+    "data",
+    [
+        {
+            "dep": "chai@4.2.0",
+            "p_url": "http://nexus:8081/repository/1/chai/-/chai-4.2.0.tgz",
+            "tarball_name": "chai-4.2.0.tgz",
+        },
+        {
+            "dep": "assertion-error@1.1.0",
```
Comment (chmeliik) [inline tests/test_workers/test_pkg_managers/test_general_js.py:246]: it looks like we don't test any "interesting" cases - anything where the `@` and `/` replacements would be relevant

File: cachito/workers/config.py [hunk 12]
```diff
@@ -36,7 +36,7 @@ class Config(object):
     cachito_gomod_ignore_missing_gomod_file = True
     cachito_gomod_strict_vendor = False
     cachito_log_level = "INFO"
-    cachito_js_download_batch_size = 30
+    cachito_js_concurrency_limit = 5
```
Comment (chmeliik) [inline cachito/workers/config.py:39]: Missed this before - could you update the README?

**Review by chmeliik — COMMENTED** [review]
Mostly LGTM

Coverage did drop, but I think it's mostly the `async_download_binary_file()` function. If there's no way to reasonably test that one, that's OK

File: tests/test_workers/test_pkg_managers/test_general_js.py (hunk 10 above)
Comment (fepas) [reply tests/test_workers/test_pkg_managers/test_general_js.py:118]: Yes, it was an issue with `black`. There was a little miscommunication on my part, sorry :D

**Comment by fepas** [comment]
> Mostly LGTM
> 
> Coverage did drop, but I think it's only the `async_download_binary_file()` function. If there's no way to reasonably test that one, that's OK

I would need to add a library just to test this function, I think it's not worth the effort.

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

The commits could probably be squashed, but I don't mind either way

File: requirements-test.in (hunk 8 above)
Comment (MartinBasti) [reply requirements-test.in:9]: I still see this change here, please keep POSIX endlines

**Comment by MartinBasti** [comment]
Please squash commits

File: tests/test_workers/test_pkg_managers/test_general_js.py [hunk 13]
```diff
@@ -247,14 +247,20 @@ def test_download_dependencies_skip_deps(
...
+        },
     ],
 )
 def test_parse_dependency(data):
 
     proxy_repo_url = "http://nexus:8081/repository/1/"
 
     result = general_js.parse_dependency(proxy_repo_url, data["dep"])
-    assert result == [data["p_url"], data["tarball_name"]]
+    print(result)
```
[the remark this answers is not in the document]
Comment (lkolacek) [reply tests/test_workers/test_pkg_managers/test_general_js.py:262]: I believe we should remove it

**Comment by lkolacek** [comment]
Looks good. Please add back new line to `requirements-test.in` and squash commits. Thank you @fepas

**Comment by fepas** [comment]
New line add to `requirements-test.in` and commits squashed. Thank you guys! :D

File: tests/test_workers/test_pkg_managers/test_general_js.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (fepas) [reply tests/test_workers/test_pkg_managers/test_general_js.py]: Ops, fixed.

**Review by lkolacek — APPROVED** [review]
